npcx/modules/scfg.h

```diff
@@ -0,0 +1,236 @@
+/* Copyright 2020 The Chromium OS Authors. All rights reserved.
+ * Use of this source code is governed by a BSD-style license that can be
+ * found in the LICENSE file.
+ */
+
+#ifndef _NPCX_SCFG_H
+#define _NPCX_SCFG_H
+
+/******************************************************************************/
+/* System Configuration (SCFG) register definitions */
+#define NPCX_DEVCNT(base_addr)                REG8((base_addr) + 0x000)
```

Comment:
Looks like the Zephyr style is to use a device read/write where the base device address is defined in the device tree.

I just looked a random example of ADCs in the main zephyr tree under drivers. I think we should adopt that style.

Also I do think we should be making the pull request on the main Zephyr tree instead of this nuvoton_hal repository since we are adding files from scratch. I don't think we will be able to "copy" the ChromiumOS files; we can use them as an example/documentation, but ultimately I think we are going to be creating wholey new files that have new copyright and Apache 2.0 license

Reply:
> Looks like the Zephyr style is to use a device read/write where the base device address is defined in the device tree. 
> I just looked a random example of ADCs in the main zephyr tree under drivers. I think we should adopt that style.
[ML] Could you give me a specific example? We do use base device address from device tree. If you mean why we don't use structure to present the registers of the driver module, the reason is the layout of the module registers is irregular. Please refer the MIWU register definitions in miwu.h. That's our burden since the Nuvoton EC has been used for two decades. 

> Also I do think we should be making the pull request on the main Zephyr tree instead of this nuvoton_hal repository since we are adding files from scratch. 
[ML] Regarding making the pull request on the main Zephyr tree for hal driver, I don't see the example in Zephyr. We might need to proceed what we have first in case blocking the other PRs on the main Zephyr tree. Of course, this is just my personal opinion.  

> I don't think we will be able to "copy" the ChromiumOS files; we can use them as an example/documentation, but ultimately I think we are going to be creating wholey new files that have new copyright and Apache 2.0 license
[ML] So, might I change copyright to Nuvoton and Apache 2.0 license? I'm not an expert of open source license. These register definitions are also a part of Nuvoton NPCX HAL drivers and I submitted them to ChromiumOS four years ago. But I don't copy the file 'registers.h' from ChromiumOS directly. The filename and layout of new files follow our hal drivers.

Reply:
Hi @jettr:
I think I got your point. Do you mean put register definitions in driver implementation itself? (https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/adc/adc_lmp90xxx.c#L29) I had thought about it before. There're two disadvantages that make driver maintenance is not easy. 
1. I have mentioned in the previous comment, we cannot use structure to present the registers of the driver module. We need to put lots register definitions in driver implementation file.
2. The register layout varies in different chip series. We need to put lots CONFIG_SOC_SERIES_NPCX7/9/N to distinguish the correct register layout. 

That's why I abstract register definitions from driver implementation to different header files. I can move them from hal_nuvoton to soc folder. But I don't see such example on Zephyr project. It might violate the rule of it.